### DISCUSSION TITLE
Multiple Databases

### DISCUSSION THREAD

**Discussion by LimpidCrypto** [loco-rs/loco#464] [Q&A] [answered]
I have a SaaS1 with a Database hosted locally using Docker. I also have another SaaS2 which is pretty similar to the SaaS1. It also has a Database.
I would need SaaS1 to be able to connect to the database of SaaS2. What would be a good way to integrate the second database connection into SaaS1, using the loco project structure?

**Comment by kaplanelad** [comment, marked as answer]
Hey @LimpidCrypto
In your `APP` hooks, navigate to the `after_routes` function and implementing the following example:
```rs
   async fn after_routes(router: axum::Router, _ctx: &AppContext) -> Result<axum::Router> {
        let saas2_db_config = config::Database {
           // ... (fill in your configuration details)
        };

        let saas2_db = db::connect(&saas_db_config).await?;
        Ok(router.layer(Extension(saas2_db)))
    }
```

In this illustration, we use a local DB configuration and integrate the database connection as an Axum extension.

Now, in your controller, retrieve this instance using the following example:
```rs
async fn example(
    _auth: auth::JWT,
    State(_ctx): State<AppContext>,
    Extension(saas2_db): Extension<DatabaseConnection>,
) -> Result<()> {
   ...
}
```
Define the extension in the controller's schema function, and you'll be able to seamlessly utilize the secondary database.

**Reply by LimpidCrypto** [reply]
Hi @kaplanelad ,
thank you for your explanation. SaaS2 is actually for managing the user database. SaaS1 needs to utilize it. Here is how I implemented it:
```rust
async fn after_routes(router: axum::Router, _ctx: &AppContext) -> Result<axum::Router> {
    let uri: &str = env!("USER_DATABASE_URL"); // URI to SaaS2 database
    dbg!(uri);
    let user_db_config = Database {
        uri: uri.to_string(),
        enable_logging: false,
        min_connections: 1,
        max_connections: 1,
        connect_timeout: 500,
        idle_timeout: 500,
        auto_migrate: false,
        dangerously_truncate: false,
        dangerously_recreate: false,
    };
    let user_db = db::connect(&user_db_config).await?;

    Ok(router.layer(Extension(user_db)))
}
```
Now I want to use the `DatebaseConnection` like so:
```rust
pub async fn list(
    State(ctx): State<AppContext>,
    auth: auth::JWT,
    Extension(user_db): Extension<DatabaseConnection>,
) -> Result<Json<Vec<Model>>> {
    dbg!(auth.claims.pid.clone());
    let user = users::Model::find_by_pid(&user_db, &auth.claims.pid).await?;
    dbg!(user);
    format::json(Entity::find().all(&ctx.db).await?)
}
```
I've imported `users::Model::find_by_pid` from the SaaS2 project. The `pid` returned by `dbg!(auth.claims.pid.clone());` exists in the SaaS2 database. But I still get the following error message:
```
2024-02-28T11:16:58.529198Z ERROR http-request: loco_rs::controller: Model(EntityNotFound) http.method=GET http.uri=/api/v1/notes http.version=HTTP/1.1 http.user_agent=PostmanRuntime/7.36.3 environment=development request_id=c44648d7-0d49-49cf-8b0c-b9ef3c62c9ae
```
Do you have any suggestions?

**Reply by LimpidCrypto** [reply]
`list` is part of the `notes` starter controller

**Reply by kaplanelad** [reply]
It looks like the given error that the connection succeeded, but the record is not found.  
So maybe you don't have the user in the second DB?

**Reply by LimpidCrypto** [reply]
Yes I thought so too but the `pid` printed by the `dbg!(auth.claims.pid.clone());` exists:
![image](https://github.com/loco-rs/loco/assets/97235361/a64debaa-148c-4bfd-9000-bf92880f16de)

**Reply by LimpidCrypto** [reply]
I inserted the user of SaaS2 into the database of SaaS1 and now it I get the notes list. But of course I want to get the user of the SaaS2 database. So it I get the wrong database connection in the controller method

**Reply by LimpidCrypto** [reply]
Oh forget about it. During testing I changed the `env!("USER_DATABASE_URL")` which points to SaaS2 database to `env!("DATABASE_URL")` which points to SaaS1 database. Your solustion works just fine. Thanks for your help :)
